Fill problematic values when creating the skycoord objects

gbrammer opened the pull request:
This PR fixes a few problematic values that can affect generating a robust `skycoord` property:

1. Enforce minimum parallax of 1 micro-arcsec when computing the `distance` property
2. Fill NaN `radial_velocity` values with zeros
3. Set the `obstime` property based on the `ref_epoch` column in the GAIA ctalogs.  

With these fixes, the `~astropy.coordinates.SkyCoord.apply_space_motion` method should work for propagating the GAIA proper motions to a given observation epoch.

adrn commented:
Thanks! I'll have a look at this today.

adrn commented on pyia/data.py:
```diff
@@ -227,7 +227,11 @@ def pm(self):
     def distance(self):
         """Assumes 1/parallax. Has shape `(nrows,)`"""
-        return coord.Distance(parallax=self.parallax)
+        # Clip to positive values (1e-3 mas = 1 Mpc)
```
I think I would prefer to default to `NaN` rather than a hard-coded clip value. But way to make us both happy could be to add a `get_distance()` method that looks something like this (I didn't test this):
```python
@u.quantity_input(fill_value=u.mas, clip_value=u.mas, 
                  equivalencies=u.parallax())
def get_distance(self, fill_value=None, clip_value=None):
    plx = self.parallax.copy()

    if clip_value is not None and fill_value is None:
        raise ValueError("If specifying a clip value, you must also")

    if clip_value is not None:
        plx[plx > clip_value.to(plx.unit, 
                                u.parallax())] = fill_value.to(plx.unit, 
                                                               u.parallax())

    if fill_value is not None:
        plx[plx <= 0] = fill_value.to(plx.unit, u.parallax())
        
    return coord.Distance(parallax=plx)

@property
def distance(self):
    """Assumes 1/parallax. Has shape `(nrows,)`. 

    This attribute will raise an error when  there are negative or zero parallax
    values. For more flexible retrieval of distance values and auto-filling bad
    values, use the .get_distance() method."""
    return self.get_distance()
```

adrn replied:
Oh, also, `get_distance()` would need a docstring.

adrn commented on pyia/data.py:
```diff
@@ -427,8 +432,18 @@ def get_skycoord(self, distance=None, radial_velocity=None):
+            # Fill nans with zeros
+            fill = ~np.isfinite(self.radial_velocity)
+            rv = self.radial_velocity*1
+            rv[fill] = 0*u.km/u.s
```
I think I'd like to go the same route with `.radial_velocity` as with distance: add a `get_radial_velocity()` method with a `fill_value` argument.

adrn replied:
`get_skycoord()` could then be called like:
```python
g.get_skycoord(distance=g.get_distance(fill_value=...),
               radial_velocity=g.get_radial_velocity(fill_value=...))
```
when there are nan or bad values in there.

adrn commented on pyia/data.py:
```diff
@@ -427,8 +432,18 @@ def get_skycoord(self, distance=None, radial_velocity=None):
+            kw['radial_velocity'] = rv
+            
+        # Reference epoch (2015.5 for DR2, could change in future)
+        obstime = astropy.time.Time(self.ref_epoch.value,
```
I like the idea of having this! But I sometimes download data without this column. Maybe we can add a `default_ref_epoch` argument to the initializer that looks something like:
```python
def __init__(self, data, ref_epoch='latest', **kwargs):
    if ref_epoch == 'latest':
        # DR2, for now
        ref_epoch = Time(2015.5, format='decimalyear')
```

gbrammer commented:
@adrn: thanks for the comments, they all make the code cleaner and my suggestions more transparent.  

I committed a change to add the "get_distance" method as suggested, though I just implemented it with a single `min_parallax` keyword.

I took out the radial velocity things since one could already ignore radial velocities by calling with `get_skycoord(radial_velocity=False)`.  

I added the ref_epoch option to `get_skycoord` rather than `__init__`, which is only applied if `self.data` is missing the "ref_epoch" column.

adrn commented on pyia/data.py:
```diff
@@ -223,12 +226,36 @@ def pm(self):
         return np.vstack((self.pmra.value, self.pmdec.to(_u).value)).T * _u
+    
+    @u.quantity_input(min_parallax=u.mas, equivalencies=u.parallax())
+    def get_distance(self, min_parallax=None, allow_negative=True):
```
I would prefer to keep the default as False, but I do like the idea of adding this argument!
```suggestion
    def get_distance(self, min_parallax=None, allow_negative=False):
```

adrn commented:
Thanks @gbrammer!
